### PR TITLE
Redoing header PR DR-2590, DR-2584

### PR DESCRIPTION
**Pull request by 7emansell**
## Ticket:

JIRA tickets [2584](https://jira.nypl.org/browse/DR-2584) and [2590](https://jira.nypl.org/browse/DR-2590)

## This PR does the following:
- Adds the sticky header with placeholders for header components
- Adds the logo

## How has this been tested?

- Does not include component tests since they will likely be outdated once search is added

### Checklist:

- [x] I have added relevant accessibility documentation for this pull request.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by EdwinGuzman** [comment]
Uh yeah what is going on and now you have merge conflicts.

**Comment by 7emansell** [comment]
Will deal with merge conflicts later– they're due to the premature merge/revert

File: src/utils/useStickyMargin.ts [hunk 1]
```diff
@@ -0,0 +1,103 @@
+import { useEffect } from "react";
+
+/* Prevents focused elements from being obscured by sticky header. */
+export function useStickyMargin() {
+  useEffect(() => {
```
Comment (7emansell) [inline src/utils/useStickyMargin.ts:5]: Line by line this code is pretty incomprehensible though I know what it's doing overall– this is a Typescript refactor of the idea presented in this article (that Clare suggested): https://www.tpgi.com/prevent-focused-elements-from-being-obscured-by-sticky-headers/. Working on trying to clarify it for myself.

File: src/utils/useStickyMargin.ts [hunk 2]
```diff
@@ -0,0 +1,103 @@
+import { useEffect } from "react";
+
+/* Prevents focused elements from being obscured by sticky header. */
+export function useStickyMargin() {
```
Comment (7emansell) [inline src/utils/useStickyMargin.ts:4]: Line by line this code is pretty incomprehensible though I know what it's doing overall– this is a Typescript refactor of the idea presented in this article (that Clare suggested): https://www.tpgi.com/prevent-focused-elements-from-being-obscured-by-sticky-headers/. Working on trying to clarify it for myself.

File: __tests__/homepage.test.tsx [hunk 3]
```diff
@@ -2,9 +2,15 @@ import { render } from "@testing-library/react";
...
 import { props } from "./data/homepageProps";
+class ResizeObserver {
+  observe() {}
+  unobserve() {}
+  disconnect() {}
+}
 
 describe("Homepage Accessibility", () => {
   it("passes axe accessibility test", async () => {
+    window.ResizeObserver = ResizeObserver;
```
Comment (EdwinGuzman) [inline __tests__/homepage.test.tsx:13]: Did this give you an error before you added this? Do you remember what it was?

File: src/utils/useScrolled.ts [hunk 4]
```diff
@@ -1,14 +1,14 @@
...
   const [isScrolled, setIsScrolled] = useState(true);
 
   useEffect(() => {
     function handleScroll() {
-      const header = document.getElementById("header");
-      if (header) {
-        const offset = header.offsetHeight;
+      const element = document.getElementById(elementId);
+      if (element) {
+        const offset = element.offsetHeight + 300;
```
Comment (EdwinGuzman) [inline src/utils/useScrolled.ts:11]: Can you make 300 a variable such as `DELAY_STICKY_TRANSITION` (based on the comments in the ticket). Just to keep a history of where this number is coming.

File: src/utils/useStickyMargin.ts [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+        applicable = !sticky.header.contains(focused);
+      }
+      if (applicable && sticky.selector) {
+        applicable = focused.matches(sticky.selector);
+      }
+
+      if (applicable) {
+        const edge = {
+          header: sticky.header
+            ? sticky.header.getBoundingClientRect().bottom + sticky.offset!
```
Comment (EdwinGuzman) [inline src/utils/useStickyMargin.ts:25]: Is the exclamation point at the end a typo?

**Review by EdwinGuzman — CHANGES_REQUESTED** [review]
Added comments for minor updates. How do you feel about moving the two hooks into a new directory `src/hooks`? They are utility functions but they follow the react pattern for hooks and naming convention so they can be separated out.

I agree, the code from the resource Clare gave you is doing a lot and the typescript/react conversion is great. I'm thinking to figure out if it all needs to be in the `useEffect` or if some can be defined outside, but it doesn't seem that way. I'd just be concern about having to declare and check for the DOM element on every render but, at least for now, there are no signs of inefficiency from the browser (e.g. being laggy). 

I do want to take a closer look at the `useStickyMargin` code and am interested in seeing comments in the code of what you figured out/clarified.

File: __tests__/homepage.test.tsx (hunk 3 above)
Comment (7emansell) [reply __tests__/homepage.test.tsx:13]: `ReferenceError: ResizeObserver is not defined`

File: src/utils/useStickyMargin.ts (hunk 5 above)
Comment (7emansell) [reply src/utils/useStickyMargin.ts:25]: Yes :(

File: src/hooks/useStickyMargin.ts
Comment (7emansell) [inline src/hooks/useStickyMargin.ts:1]: went through line by line and very much condensed, I think it actually makes sense now. I didn't realize most of the original code was only necessary if the page had multiple sticky elements @EdwinGuzman

File: src/hooks/useStickyMargin.ts [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+      ? 0
+      : parseInt(sticky.offset as any, 10);
+
+    /** Creates a ResizeObserver instance with a callback function that calls
+     * applyStickyMargin() on the sticky object.
+     **/
+    const observer = new ResizeObserver(() => applyStickyMargin(sticky));
+
+    // Handles "focus" events.
+    const applyStickyMarginWithEvent = (ev) => {
```
Comment (EdwinGuzman) [inline src/hooks/useStickyMargin.ts:62]: Nitpick: it's somewhat conventional to prefix unused function arguments with an underscore. `_event` is a better argument name. I tried it and I noticed the linter doesn't pick up that it's an "unused var". Not a big deal.

**Review by EdwinGuzman — APPROVED** [review]
Nice job on this!

File: src/hooks/useScrolled.ts [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+        const offset = element.offsetHeight + DELAY_STICKY_TRANSITION;
+        if (window.scrollY > offset) {
+          setIsScrolled(false);
+        } else {
+          setIsScrolled(true);
+        }
+      }
+    }
+    function handleScreenSize() {
+      const newDelay = window.innerWidth < 768 ? 740 : 400;
```
Comment (EdwinGuzman) [inline src/hooks/useScrolled.ts:22]: This should also be 600. Otherwise, update looks good.

File: __tests__/homepage.test.tsx (hunk 3 above)
Comment (EdwinGuzman) [reply __tests__/homepage.test.tsx:13]: I encountered this same issue in a DS PR I am working on. It affected multiple tests. Eventually, when new pages are added, this error will most likely come up again.

I did a global fix by adding your code (thanks!) into the DS [`src/__tests__/setup.ts`](https://github.com/NYPL/nypl-design-system/pull/1448/files#diff-a68a2df534784b96b4ce1a9d4d9580bc33679e819076557c07a191d2b0d5dc15) file. You can do the same by moving 
```
class ResizeObserver {
  observe() {}
  unobserve() {}
  disconnect() {}
}

window.ResizeObserver = ResizeObserver;
```

into the repo's `./jest.setup.ts` file.

**Comment by 7emansell** [comment]
See conversations and previous approvals on https://github.com/NYPL/digital-collections/pull/28, merging into main!

**Pull request merged**